x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulate.ts

```diff
@@ -0,0 +1,162 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { z } from '@kbn/zod';
+import { notFound, internal } from '@hapi/boom';
+import {
+  FieldDefinitionConfig,
+  ProcessingDefinition,
+  getProcessorType,
+  processingDefinitionSchema,
+} from '@kbn/streams-schema';
+import { calculateObjectDiff, flattenObject } from '@kbn/object-utils';
+import { get } from 'lodash';
+import {
+  IngestSimulateResponse,
+  IngestSimulateSimulateDocumentResult,
+} from '@elastic/elasticsearch/lib/api/types';
+import { createServerRoute } from '../../create_server_route';
+import { DefinitionNotFound } from '../../../lib/streams/errors';
+import { checkAccess } from '../../../lib/streams/stream_crud';
+import { conditionToPainless } from '../../../lib/streams/helpers/condition_to_painless';
+
+export const simulateProcessorRoute = createServerRoute({
+  endpoint: 'POST /api/streams/{id}/processing/_simulate',
+  options: {
+    access: 'internal',
+  },
+  security: {
+    authz: {
+      enabled: false,
+      reason:
+        'This API delegates security to the currently logged in user and their Elasticsearch permissions.',
+    },
+  },
+  params: z.object({
+    path: z.object({ id: z.string() }),
+    body: z.object({
+      processing: z.array(processingDefinitionSchema),
+      documents: z.array(z.record(z.unknown())),
+    }),
+  }),
+  handler: async ({ params, request, getScopedClients }) => {
+    try {
+      const { scopedClusterClient } = await getScopedClients({ request });
+
+      const hasAccess = await checkAccess({ id: params.path.id, scopedClusterClient });
+      if (!hasAccess) {
+        throw new DefinitionNotFound(`Stream definition for ${params.path.id} not found.`);
+      }
+      // Normalize processing definition to pipeline processors
+      const processors = normalizeProcessing(params.body.processing);
+      // Convert input documents to ingest simulation format
+      const docs = params.body.documents.map((doc) => ({ _source: doc }));
+
+      const simulationResult = await scopedClusterClient.asCurrentUser.ingest.simulate({
+        verbose: true,
+        pipeline: { processors },
+        docs,
+      });
+
+      const documents = computeSimulationDocuments(simulationResult, docs);
+      const detectedFields = computeDetectedFields(simulationResult, docs);
+      const successRate = computeSuccessRate(simulationResult);
+      const failureRate = 1 - successRate;
+
+      return {
+        documents,
+        success_rate: parseFloat(successRate.toFixed(2)),
+        failure_rate: parseFloat(failureRate.toFixed(2)),
+        detected_fields: detectedFields,
+      };
+    } catch (e) {
+      if (e instanceof DefinitionNotFound) {
+        throw notFound(e);
+      }
+
+      throw internal(e);
+    }
+  },
+});
+
+const normalizeProcessing = (processing: ProcessingDefinition[]) => {
```

Comment:
This looks very similar to https://github.com/elastic/kibana/blob/68d4cc6806fb5d7cc932d007ebe18458b6848110/x-pack/solutions/observability/plugins/streams/server/lib/streams/ingest_pipelines/generate_ingest_pipeline.ts#L17

should we make it one function?

Reply:
Addressed it [refactor(streams): extract formatToIngestProcessors helper](https://github.com/elastic/kibana/pull/204793/commits/637e4783bcd57def7f6fbf24410e32d58833d2bf)